File: Leetcode/547/547.go
const N int = 205

var (
	father [N]int
	Size[N]int
	ans int
)

func dsuInit() {
	for i := 0; i < N; i++ {
		father[i] = i
		Size[i] = 1
	}
}

func find(x int) int {
	if father[x] != x {
		father[x] = find(father[x])
	}
	return father[x]
}

func merge(x, y int) {
	x = find(x)
	y = find(y)
	if x == y {
		return
	}
	ans--
	if Size[x] > Size[y] {
		x,y = y,x
	}
	Size[y] += Size[x]
	father[x] = y
}

func findCircleNum(isConnected [][]int) int {
	dsuInit()
	n := len(isConnected)
	ans = n
	for i := 1; i <= n ; i++ {
		for j := 1; j <= n; j++ {
			if isConnected[i-1][j-1] == 1 {
				merge(i, j)
			}
		}
	}
	return ans
}

File: Leetcode/547/547.cpp
namespace DSU {
    vector<int> father;
    vector<int> Size;
    void init(int n) {
        vector<int>().swap(father);
        vector<int>().swap(Size);
        father.resize(n + 5);
        iota(father.begin(), father.end(), 0);
        Size.resize(n + 5, 1);
    }
    int find(int x) {
        if (father[x] != x) father[x] = find(father[x]);
        return father[x];
    }
    void merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) return;
        if (Size[x] > Size[y]) swap(x, y);
        Size[y] += Size[x];
        father[x] = y;
    }
}

class Solution {
public:
    int findCircleNum(vector<vector<int>>& M) {
        int n = (int)M.size();
        DSU::init(n);
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (M.at(i).at(j) == 1) {
                    DSU::merge(i, j);
                }
            }
        }
        constexpr int MAXN = 205;
        bitset<MAXN> groups;
        for (int i = 0; i < n; ++i) {
             groups[DSU::find(i)] = true;
        }
        return groups.count();
    }
};